rust/kernel/net/socket/flags.rs

```diff
@@ -0,0 +1,467 @@
+        match val {
+            bindings::SOCK_NOSPACE => Ok(SocketFlag::NoSpace),
+            bindings::SOCK_PASSCRED => Ok(SocketFlag::PassCred),
+            bindings::SOCK_PASSSEC => Ok(SocketFlag::PassSecurity),
+            bindings::SOCK_SUPPORT_ZC => Ok(SocketFlag::SupportZeroCopy),
+            bindings::SOCK_CUSTOM_SOCKOPT => Ok(SocketFlag::CustomSockOpt),
+            bindings::SOCK_PASSPIDFD => Ok(SocketFlag::PassPidFd),
+            _ => Err(()),
+        }
+    }
+}
+
+/// Flags associated with a received message.
+///
+/// Represents the flag contained in the `msg_flags` field of a `msghdr` struct.
+#[derive(Debug, Copy, Clone)]
+pub enum MessageFlag {
+    /// End of record.
+    Eor = bindings::MSG_EOR as isize,
+    /// Trailing portion of the message is discarded.
+    Trunc = bindings::MSG_TRUNC as isize,
+    /// Control data was discarded due to lack of space.
+    Ctrunc = bindings::MSG_CTRUNC as isize,
+    /// Out-of-band data was received.
+    Oob = bindings::MSG_OOB as isize,
+    /// An error was received instead of data.
+    ErrQueue = bindings::MSG_ERRQUEUE as isize,
+}
+
+impl From<MessageFlag> for isize {
+    fn from(value: MessageFlag) -> Self {
+        value as isize
+    }
+}
+
+impl TryFrom<isize> for MessageFlag {
+    type Error = ();
+
+    fn try_from(value: isize) -> Result<Self, Self::Error> {
+        let val = value as u32;
+        match val {
+            bindings::MSG_EOR => Ok(MessageFlag::Eor),
+            bindings::MSG_TRUNC => Ok(MessageFlag::Trunc),
+            bindings::MSG_CTRUNC => Ok(MessageFlag::Ctrunc),
+            bindings::MSG_OOB => Ok(MessageFlag::Oob),
+            bindings::MSG_ERRQUEUE => Ok(MessageFlag::ErrQueue),
+            _ => Err(()),
+        }
+    }
+}
+
+impl Flag for MessageFlag {}
+
+/// Structure representing a set of flags.
+///
+/// This structure is used to represent a set of flags, such as the flags passed to `send` or `recv`.
+/// It is generic over the type of flag that it contains.
+///
+/// # Invariants
+/// The value of the flags must be a valid combination of the flags that it contains.
+///
+/// This means that the value must be the bitwise OR of the values of the flags, and that it
+/// must be possible to retrieve the value of the flags from the value.
+///
+/// # Example
+/// ```
+/// use kernel::net::socket::flags::{SendFlag, FlagSet};
+///
+/// let mut flags = FlagSet::<SendFlag>::empty();
+/// flags.insert(SendFlag::DontWait);
+/// flags.insert(SendFlag::More);
+/// assert!(flags.contains(SendFlag::DontWait));
+/// assert!(flags.contains(SendFlag::More));
+/// flags.clear();
+/// assert_eq!(flags.value(), 0);
+///
+/// flags = FlagSet::<SendFlag>::from(SendFlag::More);
+/// flags |= SendFlag::DontWait;
+/// assert!(flags.contains(SendFlag::DontWait));
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct FlagSet<T: Flag> {
+    value: isize,
+    _phantom: core::marker::PhantomData<T>,
```

Comment:
Is  `isize` correct here? This would be `ssize_t` in C, and it looks like every call to `.value()` is being casted.

We might want to rethink `FlagSet` a bit anyway